src/vmcu/services/communications/spi.py

```diff
@@ -18,41 +21,84 @@ class SPIPeripheral(ABC):
     Messages are handled using queues and threads, one for transmission and other for reception.
-    '''
+    """
+
     @abstractmethod
-    def _send(self): pass
-    
+    def _send(self):
+        pass
+
     @abstractmethod
-    def _recv(self): pass
+    def _recv(self):
+        pass
 
-    def __init__(self, ip, port): 
-        self.socket = socket.socket(type=socket.SOCK_DGRAM) # Socket used to send and receive SPI data
+    def __init__(self, ip, port):
+        self.socket = socket.socket(
+            type=socket.SOCK_DGRAM
+        )  # Socket used to send and receive SPI data
         self.socket.bind((ip, port))
 
-        self.transmission_queue = queue.Queue() # Queue used to store messages that should be sended
-        self.reception_queue = queue.Queue() # Queue used to store received messages
-        self.transmission_thread = Thread(target=self._send , args=("Transmitter")) # Thread that manages data transmission
-        self.reception_thread = Thread(target=self._recv , args=("Receiver")) # Thread that manages data reception
+        self.transmission_queue = (
+            queue.Queue()
+        )  # Queue used to store messages that should be sended
+        self.reception_queue = queue.Queue()  # Queue used to store received messages
+        self.transmission_thread = Thread(
+            target=self._send, args=("Transmitter")
+        )  # Thread that manages data transmission
+        self.reception_thread = Thread(
+            target=self._recv, args=("Receiver")
+        )  # Thread that manages data reception
 
         self.transmission_thread.start()
         self.reception_thread.start()
 
-    def receive(self) -> bytes: 
+    def receive(self) -> bytes:
         return self.reception_queue.get
-    
+
     @abstractmethod
-    def transmit(self): pass
+    def transmit(self):
+        pass
 
     def __del__(self):
         self.stop = True
         self.transmission_thread.join()
         self.reception_thread.join()
         self.socket.close()
-        
 
-class SPIMaster(SPIPeripheral): 
+    def sendInput(self, data: bytes) -> Input:
+        return SPIInput(self, data)
+    
+    def sendPacket(self, packet: SPIPacket) -> Input:
+        return SPIInput(self, packet.data)
+
+    def receiveCondition(self, data: bytes) -> Condition:
+        return SPICondition(self, data)
+    
+    def receiveCondition(self, packet: SPIPacket) -> Condition:
+        return SPICondition(self, packet.data)
+
+
+class SPIInput(Input):
+    def __init__(self, spi: SPIPeripheral, data: bytes):
+        self._data = data
+        self._spi = spi
+
+    def apply(self):
+        self._spi.transmit(self._data)
+
+
+class SPICondition(Condition):
+    def __init__(self, spi: SPIPeripheral, data: bytes):
+        self._data = data
+        self._spi = spi
+
+    def check(self):
+        if self._spi.receive() != self._data:
+            raise ConditionFailedException()
```

Comment:
You should create a custom exception that actually tells you what happend by implementing the `__str__` method